lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+	var encoderType EncoderType
+	if isEncoder, encoder := StateKeyToDeSoEncoder(stateChangeEntry.KeyBytes); isEncoder && encoder != nil {
+		// Blocks are stored in a non-DeSo-Encoder format, so we need to convert them to DeSo-Encoder format by appending metadata.
+		if encoder.GetEncoderType() == EncoderTypeBlock {
+			stateChangeEntry.EncoderBytes = AddEncoderMetadataToMsgDeSoBlockBytes(stateChangeEntry.EncoderBytes, stateChangeEntry.BlockHeight)
+		}
+
+		encoderType = encoder.GetEncoderType()
+	} else {
+		// If the value associated with the key is not an encoder, then we decode the encoder from the key itself.
+		keyEncoder, err := DecodeStateKey(stateChangeEntry.KeyBytes, stateChangeEntry.EncoderBytes)
+		if err != nil {
+			glog.Fatalf("Server._handleStateSyncerOperation: Error decoding state key: %v", err)
+		}
+		encoderType = keyEncoder.GetEncoderType()
+		stateChangeEntry.Encoder = keyEncoder
+		stateChangeEntry.EncoderBytes = nil
+	}
+
+	// Set the encoder type.
+	stateChangeEntry.EncoderType = encoderType
+
+	// Set the flush ID.
+	stateChangeEntry.FlushId = flushId
+
+	// Encode the state change entry. We encode as a byte array, so the consumer can buffer just the bytes needed
+	// to decode this entry when reading from file.
+	entryBytes := EncodeToBytes(stateChangeSyncer.BlockHeight, stateChangeEntry, false)
+	writeBytes := EncodeByteArray(entryBytes)
+
+	// Add the StateChangeEntry bytes to the queue of bytes to be written to the state change file upon Badger db flush.
+	stateChangeSyncer.addTransactionToQueue(stateChangeEntry.FlushId, writeBytes)
+}
+
+// _handleStateSyncerFlush is called when a Badger db flush takes place. It calls a helper function that takes the bytes that
+// have been cached on the StateChangeSyncer and writes them to the state change file.
+func (stateChangeSyncer *StateChangeSyncer) _handleStateSyncerFlush(event *StateSyncerFlushedEvent) {
+	stateChangeSyncer.StateSyncerMutex.Lock()
+	defer stateChangeSyncer.StateSyncerMutex.Unlock()
+
+	if event.IsMempoolFlush {
+		// If this is a mempool flush, make sure a block hasn't mined since the mempool entries were added to queue.
+		// If not, reset the mempool maps and file, and start from scratch. The consumer will revert the mempool transactions
+		// it currently has and sync from scratch.
+		if stateChangeSyncer.BlockSyncFlushId != event.BlockSyncFlushId {
+			stateChangeSyncer.ResetMempool()
+			return
+		}
+
+		// Check to see if any of the keys in the mempool key set are not in the mempool key value map.
+		// If so, reset the mempool key value map, mempool key set, and mempool file, start from scratch.
+		for key, _ := range stateChangeSyncer.MempoolKeyValueMap {
+			// If any of the keys that the mempool is currently tracking weren't included in the flush, the state syncer
+			// mempool is bad and needs to be reset.
+			if _, ok := stateChangeSyncer.MempoolFlushKeySet[key]; !ok {
+				stateChangeSyncer.ResetMempool()
```

Comment:
So I've been trying really hard to figure out how MempoolKeyValueMap could ever be different from MempoolFlushKeySet. Right now, it seems like they'd always have the exact same stuff in them, and thus this inner thing would never get triggered. Am I missing something?

Reply:
`MempoolFlushKeySet` is reset every mempool flush, whereas `MempoolKeyValueMap` is reset every time a block mining occurs. The purpose of this is to detect if any entries being tracked in the mempool by state syncer become ejected. The `MempoolKeyValueMap` represents everything that has been flushed to the mempool since the last block, while the `MempoolFlushKeySet` represents only things from this latest flush. If something is in the `MempoolKeyValueMap` but is missing from `MempoolFlushKeySet`, that would indicate an entry has been ejected from the mempool since we started tracking it. The course of action here is to reset the state syncer mempool, so that the consumer can get a fresh view of the mempool.

I'll update the documentation around these fields to better represent and explain this.

Reply:
OK I get this fully now and it's a clever solution. Basically, every time we iterate over the mempool, we need to make sure we're flushing a superset of what we flushed in the past. If that's the case then it means writing only the new stuff is safe (and this is what we do in a different loop somewhere else). Otherwise, we need to reset the mempool and just write the whole thing from scratch.